ext/native/thin3d/VulkanRenderManager.cpp

```diff
@@ -0,0 +1,1328 @@
+	_dbg_assert_msg_(G3D, dstRect.offset.y + dstRect.extent.height <= (uint32_t)dst->height, "dstrect offset y + extent > height");
+
+	VKRStep *step = new VKRStep{ VKRStepType::BLIT };
+
+	step->blit.aspectMask = aspectMask;
+	step->blit.src = src;
+	step->blit.srcRect = srcRect;
+	step->blit.dst = dst;
+	step->blit.dstRect = dstRect;
+	step->blit.filter = filter;
+
+	std::unique_lock<std::mutex> lock(mutex_);
+	steps_.push_back(step);
+	curRenderStep_ = nullptr;
+	curFramebuffer_ = VK_NULL_HANDLE;
+}
+
+VkImageView VulkanRenderManager::BindFramebufferAsTexture(VKRFramebuffer *fb, int binding, int aspectBit, int attachment) {
+	// Should just mark the dependency and return the image.
+	for (int i = (int)steps_.size() - 1; i >= 0; i--) {
+		if (steps_[i]->stepType == VKRStepType::RENDER && steps_[i]->render.framebuffer == fb) {
+			// If this framebuffer was rendered to earlier in this frame, make sure to pre-transition it to the correct layout.
+			if (steps_[i]->render.finalColorLayout == VK_IMAGE_LAYOUT_UNDEFINED) {
+				steps_[i]->render.finalColorLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+				break;
+			}
+			else {
+				// May need to shadow the framebuffer if we re-order passes later.
+			}
+		}
+	}
+
+	curRenderStep_->preTransitions.push_back({ fb, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL });
+	return fb->color.imageView;
+}
+
+void VulkanRenderManager::Flush() {
+	curRenderStep_ = nullptr;
+	curFramebuffer_ = VK_NULL_HANDLE;
+	int curFrame = vulkan_->GetCurFrame();
+	FrameData &frameData = frameData_[curFrame];
+	if (frameData.hasInitCommands) {
+		vkEndCommandBuffer(frameData.initCmd);
+	}
+	if (!useThread) {
+		frameData.steps = std::move(steps_);
+		Run(curFrame);
+	} else {
+		std::unique_lock<std::mutex> lock(frameData.pull_mutex);
+		VLOG("PUSH: Frame[%d].readyForRun = true", curFrame);
+		frameData.steps = std::move(steps_);
+		frameData.readyForRun = true;
+		frameData.pull_condVar.notify_all();
+	}
+	vulkan_->EndFrame();
+}
+
+void VulkanRenderManager::Run(int frame) {
+	FrameData &frameData = frameData_[frame];
+	auto &stepsOnThread_ = frameData_[frame].steps;
+	VkDevice device = vulkan_->GetDevice();
+
+	uint32_t curSwapchainImage = 0;
+	// Get the index of the next available swapchain image, and a semaphore to block command buffer execution on.
+	// Now, I wonder if we should do this early in the frame or late? Right now we do it early, which should be fine.
+	VkResult res = vkAcquireNextImageKHR(device, vulkan_->GetSwapchain(), UINT64_MAX, acquireSemaphore_, (VkFence)VK_NULL_HANDLE, &curSwapchainImage);
+	assert(res == VK_SUCCESS);
+
+	VkCommandBuffer cmd = frameData.mainCmd;
+
+	VkCommandBufferBeginInfo begin = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO };
+	begin.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
+	begin.pInheritanceInfo = nullptr;
+	res = vkBeginCommandBuffer(cmd, &begin);
+	assert(res == VK_SUCCESS);
+
+	// TODO: Deal with the VK_SUBOPTIMAL_KHR and VK_ERROR_OUT_OF_DATE_KHR
```

Comment:
This TODO has journeyed a little ways from what it once referenced, I think?  Also now there is an extra assert here.

-[Unknown]